### PR TITLE
Handled Invalid LDAP configuration Exception and showing appropriate error message(#1186).

### PR DESCRIPTION
**Pull request by juhijariwala**

### CODE REVIEW

**Comment by juhijariwala** [comment]
We thought [`BasicAuthenticationFilter`](https://github.com/gocd/gocd/blob/80a7bf472ea7a6ef66b22374b6c4580d97d02f20/server/webapp/WEB-INF/applicationContext-acegi-security.xml#L133-L138) is the best place to handle this exception because it [knows about](https://github.com/gocd/gocd/blob/80a7bf472ea7a6ef66b22374b6c4580d97d02f20/server/webapp/WEB-INF/applicationContext-acegi-security.xml#L134-L136) all the [AuthenticationProviders](https://github.com/gocd/gocd/blob/80a7bf472ea7a6ef66b22374b6c4580d97d02f20/server/webapp/WEB-INF/applicationContext-acegi-security.xml#L107-L131) that can potentially raise any exceptions.

File: server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java [hunk 1]
```diff
@@ -32,17 +36,31 @@ protected Boolean initialValue() {
...
         }
     }
 
+    public void handleException(HttpServletRequest httpRequest, HttpServletResponse httpResponse, Exception e) throws IOException {
+        SecurityContext context = SecurityContextHolder.getContext();
+        String message = "There was an error authenticating you. Please check the server logs, or contact your the go administrator.";
+        httpRequest.getSession().setAttribute(AbstractProcessingFilter.SPRING_SECURITY_LAST_EXCEPTION_KEY, new RuntimeException(message));
+        httpRequest.setAttribute(SessionDenialAwareAuthenticationProcessingFilterEntryPoint.SESSION_DENIED, true);
+        context.setAuthentication(null);
+        httpResponse.sendRedirect("/go/auth/login?login_error=1");
```
Comment (arvindsv) [inline server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java]: Should this respond with a content-type aware response?

**Comment by chopraapooja** [comment]
PR is reviewed. It seems good. Ready to merge.

File: server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java [hunk 2]
```diff
@@ -32,18 +39,73 @@ protected Boolean initialValue() {
...
         }
     }
 
+    public void handleException(HttpServletRequest httpRequest, HttpServletResponse httpResponse, Exception e) throws IOException {
+        String message = "There was an error authenticating you. Please check the server logs, or contact your the go administrator.";
+        if (hasAccept(httpRequest, "text/html") || hasAccept(httpRequest, "application/xhtml")) {
+            SecurityContext context = SecurityContextHolder.getContext();
+            httpRequest.getSession().setAttribute(AbstractProcessingFilter.SPRING_SECURITY_LAST_EXCEPTION_KEY, new RuntimeException(message));
+            httpRequest.setAttribute(SessionDenialAwareAuthenticationProcessingFilterEntryPoint.SESSION_DENIED, true);
+            context.setAuthentication(null);
```
Comment (srinivasupadhya) [inline server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java]: this causes an already logged in user to be logged out (if he tries to login with invalid credentials). intentional?

File: server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java [hunk 3]
```diff
@@ -32,18 +39,73 @@ protected Boolean initialValue() {
...
             super.doFilterHttp(httpRequest, httpResponse, chain);
+        } catch (Exception e) {
+            handleException(httpRequest, httpResponse, e);
         } finally {
             isProcessingBasicAuth.set(false);
         }
     }
 
+    public void handleException(HttpServletRequest httpRequest, HttpServletResponse httpResponse, Exception e) throws IOException {
+        String message = "There was an error authenticating you. Please check the server logs, or contact your the go administrator.";
```
Comment (srinivasupadhya) [inline server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java]: should go to `localize.xml`

File: server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java [hunk 4]
```diff
@@ -32,18 +39,73 @@ protected Boolean initialValue() {
...
+        if (hasAccept(httpRequest, "text/html") || hasAccept(httpRequest, "application/xhtml")) {
+            SecurityContext context = SecurityContextHolder.getContext();
+            httpRequest.getSession().setAttribute(AbstractProcessingFilter.SPRING_SECURITY_LAST_EXCEPTION_KEY, new RuntimeException(message));
+            httpRequest.setAttribute(SessionDenialAwareAuthenticationProcessingFilterEntryPoint.SESSION_DENIED, true);
+            context.setAuthentication(null);
+
+            httpResponse.sendRedirect("/go/auth/login?login_error=1");
+            LOG.error(e.getMessage());
+            LOG.trace(e.getMessage(), e);
```
Comment (srinivasupadhya) [inline server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java]: are 2 log statements required?

**Pull request merged**